base/loading.jl

```diff
@@ -1626,29 +1631,31 @@ function get_preferences(uuid::UUID)
         end
 
         # Collect all dictionaries from the current point in the load path, then merge them in
-        dicts = collect_preferences!(project_toml, uuid)
+        dicts = collect_preferences(project_toml, uuid)
         merged_prefs = recursive_prefs_merge(merged_prefs, dicts...)
     end
     return merged_prefs
 end
 
-function get_preferences_hash(uuid::UUID, prefs_list::Vector{String})
+function get_preferences_hash(uuid::Union{UUID, Nothing}, prefs_list::Vector{String})
     # Start from the "null" hash
-    h = get_preferences_hash(nothing, prefs_list)
+    h = UInt64(0x6e65726566657250)
+    uuid === nothing && return h
 
     # Load the preferences
     prefs = get_preferences(uuid)
 
     # Walk through each name that's called out as a compile-time preference
     for name in prefs_list
-        if haskey(prefs, name)
-            h = hash(prefs[name], h)
+        prefs_name = get(prefs, name, nothing)::Union{String, Nothing}
+        if prefs_name !== nothing
+            h = hash(prefs_name, h)
         end
     end
     return h
 end
+
 get_preferences_hash(m::Module, prefs_list::Vector{String}) = get_preferences_hash(PkgId(m).uuid, prefs_list)
-get_preferences_hash(::Nothing, prefs_list::Vector{String}) = UInt64(0x6e65726566657250)
```

Comment:
It is quite valuable to have you comb over this after me, because I forget that I have eliminated the `nothing` sentinel value from the control flow.